### ISSUE TITLE
Implement dithering for gradients

### ISSUE THREAD

**Issue by gaaclarke** [flutter/flutter#118073]
Gradients in Flutter can sometimes experience banding.  If we implemented dithering they would look nicer.

I'm not aware of a user directly reporting this issue.  Although it takes a person with a bit of graphics savvy to be able to explain the issue.   I played around with the LinearGradient widget on a full width 4k display and it was hard for me to get a combination that was noticeable to me on my macbook.

Here is a user that was experiencing banding when using the `Paint` class: https://github.com/flutter/flutter/issues/112498#issuecomment-1371736888

I've personally run into this problem in a game I developed years ago where the designer gave me the parameters and banding appeared.

## example
<img width="1727" alt="Screen Shot 2023-01-05 at 4 52 24 PM" src="https://user-images.githubusercontent.com/30870216/210907719-4f4a1a8d-e28a-4d39-9e99-3635a26a0c74.png">

*click to maximize the image to see banding better*

## reproduction code
```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Container(
          width: double.infinity,
          decoration: const BoxDecoration(
              gradient: LinearGradient(
                  colors: <Color>[Color(0xffcccccc), Color(0xff333333)])),
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              const Text(
                'You have pushed the button this many times:',
              ),
              Text(
                '$_counter',
                style: Theme.of(context).textTheme.headlineMedium,
              ),
            ],
          ),
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
```

**Comment by jonahwilliams** [comment]
does this look better with the existing dithering flag set to true?

**Comment by gaaclarke** [comment]
> does this look better with the existing dithering flag set to true?

I don't think there is an option for dithering in `LinearGradient` unless I'm missing something.  Given the screenshots that b3nni97 gave in https://github.com/flutter/flutter/issues/112498, I'd expect Skia's dithering function to fix the issue.

**Comment by jonahwilliams** [comment]
are you referring to with impeller specifically?

**Comment by gaaclarke** [comment]
> are you referring to with impeller specifically?

No, not specifically.  We should do this in both backends.  I imagine there is a fragment shader in impeller that is doing linear gradients, it should be expanded to support dithering.

**Comment by jonahwilliams** [comment]
For the Skia backend, dithering in gradients should be enabled by setting the property on Paint

**Comment by Kiruel** [comment]
`Paint.enableDithering = true;` work well on iOS/Android. But when compile for Web I can see the banding. They are a way to force it in Web too ?

**Comment by sdegenaar** [comment]
I am seeing the problem as well. On macOS if I set the flag Paint.enableDithering = true; as previously mentioned it fixes it for macOS but stays the same on web. Can we force this or fix this to work for web as well? Really effecting my web project as not a good look. Thanks :-)

**Comment by adigladi** [comment]
Setting `Paint.enableDithering = true;` doesn't seem to work with Impeller, is this a known issue? Setting the same thing when impeller disabled works.

**Comment by kopino4** [comment]
> Setting `Paint.enableDithering = true;` doesn't seem to work with Impeller, is this a known issue? Setting the same thing with impeller disabled works.

I confirm the same as @adigladi. When setting `Paint.enableDithering = true` and define all gradient colors without opacity, The issue is resolved on Android. On iOS only if i disable impeller.

**Comment by LeoSandbox** [comment]
I am facing the same issue on iOS for my RPG game - see screenshot of a screen I'm working on inside the gradient area of my non-modal bottomsheet.
I have `Paint.enableDithering = true;` and it worked fine until Impeller came in the stable branch and was enabled by default. Don't get me wrong, Impeller is a game changer but this gradient banding needs to be patched to allow dithering.

For reference, the area in the screenshot is painted using a CustomPainter with:
`paintBackground.shader = LinearGradient(colors: [color1, color2], begin: Alignment.bottomCenter, end: Alignment.topCenter).createShader(Rect.fromLTWH(0, 0, size.width, size.height));`
            
![Screenshot 2023-07-09 at 08 01 22](https://github.com/flutter/flutter/assets/132463706/98dd899b-ec67-453a-90de-7006a3a90335)

**Comment by LeoSandbox** [comment]
Hi @adigladi @sdegenaar - can you put a thumbs up to @gaaclarke top comment dated Jan 5? That'll help prioritize the issue.

Hi @jonahwilliams - Sorry to bother you, I can contribute to code, but I’m not sure where in flutter/engine/impeller a patch should be made (which areas or files). Also I'm not familiar with Impeller's architecture but I can learn. That would be awesome if you could point me to the right direction. Thanks so much! 😊

**Comment by LeoSandbox** [comment]
Thanks for pointing me to the right direction @jonahwilliams 😊

I modified the fragment shader [linear_gradient_ssbo_fill.frag](https://github.com/flutter/engine/blob/main/impeller/entity/shaders/linear_gradient_ssbo_fill.frag) to add dithering with a bool to enable it. Changes are highlighted after the code. Let me know what you think.

```
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

precision mediump float;

#include <impeller/color.glsl>
#include <impeller/texture.glsl>
#include <impeller/types.glsl>

struct ColorPoint {
  vec4 color;
  float stop;
};

layout(std140) readonly buffer ColorData {
  ColorPoint colors[];
}
color_data;

uniform FragInfo {
  highp vec2 start_point;
  highp vec2 end_point;
  float alpha;
  float tile_mode;
  vec4 decal_border_color;
  int colors_length;
}
frag_info;

uniform bool dither;

highp in vec2 v_position;

out vec4 frag_color;

// Granularity of dither noise set to very small number 0.5 / 255.0 to ensure any shift in color due to dither noise is minimal
const highp float DITHERING_GRANULARITY = 0.001960784313725;

float random (vec2 st) {
    return fract(sin(dot(st.xy, vec2(12.9898,78.233))) * 43758.5453123);
}

void main() {
  highp vec2 start_to_end = frag_info.end_point - frag_info.start_point;
  highp vec2 start_to_position = v_position - frag_info.start_point;
  highp float t =
      dot(start_to_position, start_to_end) / dot(start_to_end, start_to_end);

  if ((t < 0.0 || t > 1.0) && frag_info.tile_mode == kTileModeDecal) {
    frag_color = frag_info.decal_border_color;
  } else {
    t = IPFloatTile(t, frag_info.tile_mode);

    for (int i = 1; i < frag_info.colors_length; i++) {
      ColorPoint prev_point = color_data.colors[i - 1];
      ColorPoint current_point = color_data.colors[i];
      if (t >= prev_point.stop && t <= current_point.stop) {
        float delta = (current_point.stop - prev_point.stop);
        if (delta < 0.001) {
          frag_color = current_point.color;
        } else {
          float ratio = (t - prev_point.stop) / delta;
          frag_color = mix(prev_point.color, current_point.color, ratio);
        }
        break;
      }
    }
  }

  if (dither) {
    frag_color += mix(-DITHERING_GRANULARITY, DITHERING_GRANULARITY, random(v_position));
  }

  frag_color = IPPremultiply(frag_color) * frag_info.alpha;
}
```

**CHANGES**

Bool input
`uniform bool dither;`

Random pattern ([source](https://thebookofshaders.com/10/))
```
float random (vec2 st) {
    return fract(sin(dot(st.xy, vec2(12.9898,78.233))) * 43758.5453123);
}
```

Adding noise to color output if dither true ([source](https://shader-tutorial.dev/advanced/color-banding-dithering/)) 
```
const highp float DITHERING_GRANULARITY = 0.001960784313725;

  if (dither) {
    frag_color += mix(-DITHERING_GRANULARITY, DITHERING_GRANULARITY, random(v_position));
  }
```

**Comment by LeoSandbox** [comment]
Is there any way I can test this locally?
I have some GLSL files in an `Impeller` folder inside my `flutter` folder on my Mac mini like `/Users/[username]/flutter/bin/cache/artifacts/engine/darwin-x64/shader_lib/impeller` but I don't see all the files that are here `https://github.com/flutter/engine/tree/main/impeller/entity/shaders`.

Looks like they may be "compiled" somehow when I install Flutter. Otherwise I'll wait for you guys.
Thanks for your time on this!

**Comment by LeoSandbox** [comment]
I just updated my answer code above, I just realized `frag_color` is a vec4 so I changed it to avoid adding noise to the alpha channel and only apply to rgb.

**Comment by Zabadam** [comment]
> float random (vec2 st) {
>     return fract(sin(dot(st.xy, vec2(12.9898,78.233))) * 43758.5453123);
> }

I know this function is present everywhere on the Internet, widley-used and easily integrated, but would a blue noise texture not give more visually pleasing results? Furthermore, using a known texture may allow predictable output (i.e. for goldens).

I am independently tackling a number of `Gradient` shortcomings, such as TileMode and GradientTransform lerping, while building a new system of shape gradients; to include dithering if it is intended to be added to vanilla `Gradient`s would be a new objective for me.

**Comment by LeoSandbox** [comment]
Yes blue noise texture is great also. There are pro and cons to both imo:

**Random Noise**

PRO: 
- For this specific dithering issue the code is ready above for testing by someone with the local engine builds and it "only" needs integration for the dither boolean
- Widely used and proven to reduce banding
- Low performance impact

CON:
- Rendering quality can be better with more modern dithering techniques

**Blue noise texture**

PRO:
- Better rendering than random noise in most cases
- Like you said, allow predictable output (i.e. for goldens)

CON:
- Depending on implementation it can have a higher performance impact, and noticing how iPhone can sometimes overheat with flutter on impeller when an image is slightly oversized or too much animations on screen, I would personally be worried if any additional performance drag is introduced.
- The code was not made yet, someone (you/me?) need to code this
- Potential lower quality when gradients are transparent and stacked with other images along with some motions (never experienced it myself but I read about it)

Considering that this issue has only **18** 👍 and is currently on page 25/500+ of all flutter issues, it is still (very) likely that it will never be fixed and dithering not implemented in near future. My 2 cents (but I'm not the owner of flutter repo in any way) is that better to fix the banding issue with the quick fix above in a matter of weeks/months, that will fit most dithering purpose, than incorporate the dithering fix to a bigger project that may or may not be part of a future flutter updates in couple of years.

Off-topic: I really like your Foil repo! I am actually using a variation of xl accelerometer code for my RPG [drakontias.com](drakontias.com) for the card collectibles in my card battle mini-game! 😊

**Comment by jonahwilliams** [comment]
FWIW you do not need to fork the engine or wait for this feature to have dithered gradients. With the fragment program API you can define your own gradients that interoplate how you want and use whatever noise approach for dithering.

Here is a repo I created that has some non-srgb gradients as an example: https://github.com/jonahwilliams/awesome_gradients

**Comment by sdegenaar** [comment]
Hi Jonah, you would not have an example using the linear_gradient_ssbo_fill.frag at all to try the dithering? 
Thanks!

**Comment by matanlurey** [comment]
I'll self-assign this issue to explore what it would take to implement this, and we can decide whether or not to land it.

Note that I'll only be exploring for _Impeller_, i.e. I won't be looking into web support for `Paint.enableDithering`.

**Comment by LeoSandbox** [comment]
Thanks so much @matanlurey for this!! Let me and @Zabadam know if you need any help.

I just saw your reply @jonahwilliams super interesting, thanks! I didn't know we could tweak shaders from flutter directly without going into the weeds of the engine. I'll re-use this code! 👀

**Comment by jonahwilliams** [comment]
You can't adjust the engine's shaders as they are precompiled, but you can provide your own via the FragmentProgram API.

**Comment by matanlurey** [comment]
Closed via https://github.com/flutter/engine/pull/44522.

Note that you'll need to be on the `master` channel to immediately see the changes. I'll be following up on https://github.com/flutter/flutter/issues/112498 on the future-proof API, but for now `Paint.enableDithering = true` will be honored for gradients on the Impeller backend.
